pkg/chunk/storage/factory.go

```diff
@@ -73,7 +75,7 @@ func (cfg *Config) Validate() error {
 }
 
 // NewStore makes the storage clients based on the configuration.
-func NewStore(cfg Config, storeCfg chunk.StoreConfig, schemaCfg chunk.SchemaConfig, limits StoreLimits) (chunk.Store, error) {
+func NewStore(cfg Config, storeCfg chunk.StoreConfig, schemaCfg chunk.SchemaConfig, limits StoreLimits, customIndexClients map[string]IndexClientFactoryFunc) (chunk.Store, error) {
```

Comment:
I'm not a big fan of passing `customIndexClients` here. An alternative approach may be:

1. Define a global variable `customIndexClients`
2. Add a function `RegisterIndexClient(name string, factory IndexClientFactoryFunc)` which stores the index client to `customIndexClients`
3. In `NewIndexClient()`, before `switch name {` we do check if `customIndexClients` contains it. If so, it will take precedence over the defaults.